cmd/cluster-replication.go

```diff
@@ -0,0 +1,968 @@
+				target.Type == madmin.ReplicationService {
+				targetARN = target.Arn
+				break
+			}
+		}
+		if targetARN == "" {
+			bucketTarget := madmin.BucketTarget{
+				SourceBucket: bucket,
+				Endpoint:     ep.Host,
+				Credentials: &madmin.Credentials{
+					AccessKey: creds.AccessKey,
+					SecretKey: creds.SecretKey,
+				},
+				TargetBucket:    bucket,
+				Secure:          ep.Scheme == "https",
+				API:             "s3v4",
+				Type:            madmin.ReplicationService,
+				Region:          "",
+				ReplicationSync: false,
+			}
+			bucketTarget.Arn = globalBucketTargetSys.getRemoteARN(bucket, &bucketTarget)
+			err := globalBucketTargetSys.SetTarget(ctx, bucket, &bucketTarget, false)
+			if err != nil {
+				fmt.Printf("%s->%s: target err: %#v\n", c.state.Name, peer.Name, err)
+				return wrapCRErr(err)
+			}
+			targets, err := globalBucketTargetSys.ListBucketTargets(ctx, bucket)
+			if err != nil {
+				return wrapCRErr(err)
+			}
+			tgtBytes, err := json.Marshal(&targets)
+			if err != nil {
+				return wrapCRErr(err)
+			}
+			if err = globalBucketMetadataSys.Update(bucket, bucketTargetsFile, tgtBytes); err != nil {
+				return wrapCRErr(err)
+			}
+			targetARN = bucketTarget.Arn
+
+			// fmt.Printf("%s->%s: target created: %#v\n", c.state.Name, peer.Name, bucketTarget)
+		}
+
+		// Create bucket replication rule to this peer.
+
+		// To add the bucket replication rule, we fetch the current
+		// server configuration, and convert it to minio-go's
+		// replication configuration type (by converting to xml and
+		// parsing it back), use minio-go's add rule function, and
+		// finally convert it back to the server type (again via xml).
+		// This is needed as there is no add-rule function in the server
+		// yet.
+		replicationConfigS, err := globalBucketMetadataSys.GetReplicationConfig(ctx, bucket)
+		if err != nil {
+			_, ok := err.(BucketReplicationConfigNotFound)
+			if !ok {
+				return wrapCRErr(err)
+			}
+		}
+		var replicationConfig replication.Config
+		if replicationConfigS != nil {
+			replCfgSBytes, err := xml.Marshal(replicationConfigS)
+			if err != nil {
+				return wrapCRErr(err)
+			}
+			err = xml.Unmarshal(replCfgSBytes, &replicationConfig)
+			if err != nil {
+				return wrapCRErr(err)
+			}
+		}
+		// FIXME: The rule-priority uniqueness check inside `.AddRule()`
+		// is supposed to generate an error only when the rule being
+		// added is for the same destination bucket as an existing rule
+		// (with the same priority) - so to migitate this, we are
+		// picking the priority randomly. If not for this bug, we could
+		// set the priority as "0" always as the destination buckets are
+		// always distinct with CR. Ref:
+		// https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-add-config.html#replication-config-min-rule-config
+		rand.Seed(time.Now().UnixNano())
+		err = replicationConfig.AddRule(replication.Options{
+			Priority:                fmt.Sprintf("%d", rand.Intn(1<<30)),
```

Comment:
Might want to add a helper method to replicationConfig to get max priority so far and use maxPriority+1 here instead of a random number

Reply:
Looking into this.

Reply:
This is done - added helper in the same file.